RGB LED Control V3.0

HossamElwahsh opened the pull request:
~Video is being uploaded (internet is slow - will get it up in less than an hour)~ Done ✅ 

#### The 'led off' state in the sequence (missing from the video) is now updated in the code.
-----

# RGB LED Brightness Control
#### ARM
#### Author: Team 1 - Sub team A
- #### Members:
    - Alaa Hisham
    - Hossam Elwahsh

## Brief
> Develop the Timer Driver and use it to control the RGB LED brightness on the Tiva-C board based on the push button press.

## Quick Links
- `Video` 👉 [Youtube](https://youtu.be/-Q7QAzkUJzY)
- `Project Hex` 👉 [LED_Control.hex](https://github.com/HossamElwahsh/RGBLedBrightnessControl-Project_TivaC-Sprints/blob/feat-brightness-control/RGB-BRIGHTNESS/Objects/LED_Control.hex)
- `Documentation` 👉 [Google Docs](https://docs.google.com/document/d/1G7hjrKdbhdpVfhLmX3FCOfq5wveIrZisSxEtebcybI0/edit?usp=sharing) 👉 [PDF](https://github.com/HossamElwahsh/RGBLedBrightnessControl-Project_TivaC-Sprints/blob/feat-brightness-control/Documents/ARM%20RGB%20LED%20Brightness%20Control%20-%20Documentation.pdf)
- `Team Backlog / Test Protocol` 👉 [Google Sheets](https://docs.google.com/spreadsheets/d/1GbQpRRLHq_ikWl7TjrVWzOoL4nGz3XgLzbhVqFADxBA/edit?usp=sharing) 👉 [Excel](https://github.com/HossamElwahsh/RGBLedBrightnessControl-Project_TivaC-Sprints/blob/feat-brightness-control/Documents/ARM%20RGB%20LED%20Brightness%20Control%20-%20Backlog%20and%20test%20protocol.xlsx) 

----
## Requirements
### Hardware Requirements
- Use the TivaC board
- Use SW1 as an input button
- Use the RGB LED
### Software Requirements
- **The RGB LED is OFF initially**
- The PWM signal has a 500ms duration
- The system has four states:
- SW1 - First press
  - The Green LED will be on with a 30% duty cycle
- SW1 - Second press
  - The Green LED will be on with a 60% duty cycle
- SW1 -Third press
  - The Green LED will be on with a 90% duty cycle
- SW1 - Fourth press will be off
  - The Green LED will be off
- On the fifth press, system state will return to state 1

### Implement your drivers
- Implement the General Purpose Timer (GPT) driver
  - GPT Driver shall support APIs like:( Initialization, starting a timer in microseconds,  milliseconds, and seconds, stopping the timer, getting elapsed time, getting remaining time, enabling interrupts, disable interrupts, and a callback.)
- Implement PWM Driver

HossamElwahsh committed:
remove: extra files

HossamElwahsh committed:
update (std): add max values to byte's sizes

HossamElwahsh committed:
update (main): add test functions to GPT

HossamElwahsh committed:
update (build): clock selection to 16MHz

HossamElwahsh committed:
create: GPT module

HossamElwahsh committed:
update: app main flow

HossamElwahsh committed:
update (gpt): add elapsed / rem APIs

HossamElwahsh committed:
refactor (app): whitespaces

HossamElwahsh committed:
update (gpt): better abstraction for callbacks

HossamElwahsh committed:
update: readme

HossamElwahsh committed:
update (app): magic numbers review

HossamElwahsh committed:
cleanup

HossamElwahsh committed:
add (doc): layered arch

HossamElwahsh committed:
update (gpt): init API (void)

HossamElwahsh committed:
doc: GPT documentation

HossamElwahsh committed:
refactor: minor fix

AalaaIbrahim committed:
add (pwm): pwm module

AalaaIbrahim committed:
update (app): pwm integration

AalaaIbrahim committed:
update (btn): gpt compatibility

AalaaIbrahim committed:
cleanup

AalaaIbrahim committed:
fix: various app bugs

AalaaIbrahim committed:
added project hex

HossamElwahsh committed:
doc: documentation files

HossamElwahsh committed:
add: documentation pdf

AalaaIbrahim committed:
add (doc): pwm flowcharts

HossamElwahsh committed:
update (readme): add video link

Hala-Abdelkader commented on RGB-BRIGHTNESS/APP/app.c:
```diff
@@ -75,56 +91,55 @@ en_app_error_t app_init(void)
             // set app state -> switching
             gl_en_app_state = SWITCHING;
+
+            switch (gl_en_app_sub_state) {
```
Switch case should have only one break

HossamElwahsh replied:
There was no proof of such rule in MISRA, and to be frank it isn't meaningful. that's the whole point of a switch() statement to have multiple options instead of doing too many if/else nesting.

Hala-Abdelkader replied:
I don't know the exact version which mentions this rule so you have two solutions to avoid doing switch and to be compliant with this MISRA rule:

1.The easy way to replace the switch case with if condition.
2.The hard way is to use Fall-through switch cases.

HossamElwahsh replied:
> I don't know the exact version which mentions this rule so you have two solutions to avoid doing switch and to be compliant with this MISRA rule:
> 
> 1.The easy way to replace the switch case with if condition. 2.The hard way is to use Fall-through switch cases.

Thanks again for the detailed input, sorry for the late reply,

We modified the code to comply with the rule you've mentioned, we opted out to option 1 (used if/else) instead
We also optimized the PWM driver which helped in optimizing the application flow, as we now removed the `led_interface.h` include from the APP file as the PWM driver is now fully controlling and managing the PWM signals.

Commit Ref: 5fb6fd8d07dada624f1ba7e4acb3bdfd9b4d11b0

Hala-Abdelkader replied:
Great work...I re-reviewed your code again and would like to thank you for your effort.

Hala-Abdelkader reviewed (changes requested):
-led.c,app.h,systick.h files are missing
-In the video, when you press the button for the fourth time, the led should turn off not repeat again the sequence

HossamElwahsh commented:
> -led.c,app.h,systick.h files are missing -In the video, when you press the button for the fourth time, the led should turn off not repeat again the sequence.

Thanks for your input,
To elaborate on the PR files, all the unchanged files are still on the main branch of this repository which includes the files you just mentioned, the PR only contains the files that have been changed from V2, to V3. 

For the fourth `completely off` state, you are right, will quickly add that to the code.

AalaaIbrahim committed:
update (pwm): reset channel pins at stop

AalaaIbrahim committed:
update (app): added led off state

AalaaIbrahim commented:
> -led.c,app.h,systick.h files are missing -In the video, when you press the button for the fourth time, the led should turn off not repeat again the sequence

Application sequence is now updated in b0b34cf42159e0d8126d94c3ff6daeb067b25dfb. Thank you.

HossamElwahsh committed:
review (app): code optimizations

- switch cases is now if/else to comply with MISRA rules
- removed LED include from app as PWM module will take care of everything related to the PWM signals including the port/pin for the LEDs

Hala-Abdelkader commented on RGB-BRIGHTNESS/APP/app.c:
```diff
@@ -74,148 +77,78 @@ en_app_error_t app_init(void)
-                gl_en_app_state  = IDLE;
-                break;
+            else {
+                gl_en_app_sub_state = SUB_STATE_0_DUTY_IN_PERCENT;
```
Well done
